pkg/mcp/sink/sink.go

```diff
@@ -0,0 +1,323 @@
+// limitations under the License.
+
+package sink
+
+import (
+	"io"
+	"sort"
+	"sync"
+
+	"github.com/gogo/protobuf/proto"
+	"github.com/gogo/protobuf/types"
+	"github.com/gogo/status"
+	"google.golang.org/grpc/codes"
+
+	mcp "istio.io/api/mcp/v1alpha1"
+	"istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/mcp/monitoring"
+)
+
+var scope = log.RegisterScope("mcp", "mcp debugging", 0)
+
+// Object contains a decoded versioned object with metadata received from the server.
+type Object struct {
+	TypeURL  string
+	Metadata *mcp.Metadata
+	Body     proto.Message
+}
+
+// changes is a collection of configuration objects of the same protobuf type.
+type Change struct {
+	Collection string
+
+	// List of resources to add/update. The interpretation of this field depends
+	// on the value of Incremental.
+	//
+	// When Incremental=True, the list only includes new/updated resources.
+	//
+	// When Incremental=False, the list includes the full list of resources.
+	// Any previously received resources not in this list should be deleted.
+	Objects []*Object
+
+	// List of deleted resources by name. The resource name corresponds to the
+	// resource's metadata name.
+	//
+	// Ignore when Incremental=false.
+	Removed []string
+
+	// When true, the set of changes represents an incremental resource update. The
+	// `Objects` is a list of added/update resources and `Removed` is a list of delete
+	// resources.
+	//
+	// When false, the set of changes represents a full-state update for the specified
+	// type. Any previous resources not included in this update should be removed.
+	Incremental bool
+}
+
+// Updater provides configuration changes in batches of the same protobuf message type.
+type Updater interface {
+	// Apply is invoked when the node receives new configuration updates
+	// from the server. The caller should return an error if any of the provided
+	// configuration resources are invalid or cannot be applied. The node will
+	// propagate errors back to the server accordingly.
+	Apply(*Change) error
+}
+
+// Stream is for sending RequestResources messages and receiving Resource messages.
+type Stream interface {
+	Send(*mcp.RequestResources) error
+	Recv() (*mcp.Resources, error)
+}
+
+type perCollectionState struct {
+	versions map[string]string
+	updated  bool
+}
+
+// Sink implements the resource sink message exchange for MCP. It can be instantiated by client and server
+// sink implementations to manage the MCP message exchange.
+type Sink struct {
+	mu    sync.Mutex
+	state map[string]*perCollectionState
+
+	nodeInfo *mcp.SinkNode
+	updater  Updater
+	journal  *recentRequestsJournal
+	metadata map[string]string
+	reporter monitoring.Reporter
+}
+
+type Options struct {
+	Collections []string
+	Updater     Updater
+	ID          string
+	Metadata    map[string]string
+	Reporter    monitoring.Reporter
+}
+
+// NewSink creates a new resource sink.
+func NewSink(options *Options) *Sink { // nolint: lll
```

Comment:
It would be nice to call this New so the caller can do `sink.New` as oppose to `sink.NewSink`, but very minor.